pkg/parser/parser.go

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Package is the set of metadata and objects in a package.
+type Package struct {
+	meta    []runtime.Object
+	objects []runtime.Object
+}
+
+// NewPackage creates a new Package.
+func NewPackage() *Package {
+	return &Package{}
+}
+
+// GetMeta gets metadata from the package.
+func (p *Package) GetMeta() []runtime.Object {
+	return p.meta
+}
+
+// GetObjects gets objects from the package.
+func (p *Package) GetObjects() []runtime.Object {
+	return p.objects
+}
+
+// Parser is a package parser.
+type Parser interface {
+	Parse(context.Context, io.ReadCloser) (*Package, error)
+}
+
+// DefaultParser is the default Parser implementation.
+type DefaultParser struct {
+	metaScheme *runtime.Scheme
+	objScheme  *runtime.Scheme
+}
+
+// New returns a new DefaultParser.
+func New(meta, obj *runtime.Scheme) *DefaultParser {
```

Comment:
What do you think about taking an interface rather than the concrete *Scheme type? Would it help with testing?

Reply:
Sounds good :) I meant to implement that but got sidetracked on how I wanted to do it. I think I'll add an `ObjectCreaterTyper` interface that wraps `runtime.ObjectCreater` and `runtime.ObjectTyper` from `runtime` as you suggested in Slack